File: dsa_leetcode_beginner/0338-counting-bits/0338-counting-bits.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<int> countBits(int n) {
    vector<int> result(n + 1, 0);
    
    if (n == 0) {
        return result;
    }

    result[1] = 1;

    for (int i = 2; i <= n; i++) {
        if (i % 2 == 0) {
            result[i] = result[i / 2];
        } else {
            result[i] = result[i / 2] + 1;
        }
    }

    return result;
}

int main() {
    int n = 5;
    vector<int> result = countBits(n);

    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    return 0;
}

File: dsa_leetcode_beginner/0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        # Initialize a list to store the result
        result = [0] * (n + 1)
        
        # If n is 0, simply return the result list (which is [0])
        if n == 0:
            return result
        
        # The number of 1's in the binary representation of 1 is 1
        result[1] = 1
        
        # Loop through numbers from 2 to n
        for i in range(2, n + 1):
            if i % 2 == 0:
                result[i] = result[i // 2]
            else: 
                result[i] = result[i // 2] + 1
        
        return result
    

#Question: https://leetcode.com/problems/counting-bits
#Blog: https://blog.unwiredlearning.com/counting-bits